File: day03/day03.py
from typing import Dict, Tuple
# import time
# start_time = time.time()

with open('input.txt', 'r') as file:
    path1, path2 = file.read().split('\n')

dr = {'U': 1, 'D': -1, 'L': 0, 'R': 0}
dc = {'R': 1, 'L': -1, 'U': 0, 'D': 0}

visited: Dict[Tuple[int,int], int]
visited = {} # maps coordinates to steps taken to get there

r = c = length = 0

for command in path1.split(','):
    op = command[0] # def not copying leo
    steps = int(command[1:])

    for step in range(steps):
        r += dr[op]
        c += dc[op]
        length += 1
        visited[(r,c)] = length

min_distance = float('inf')
r = c = length = 0

for command in path2.split(','):
    op = command[0]
    steps = int(command[1:])

    for step in range(steps):
        r += dr[op]
        c += dc[op]
        length += 1

        if (r,c) in visited:
            min_distance = min(min_distance, length + visited[(r,c)])
            print(min_distance)

# print("--- %.6f seconds ---" % (time.time() - start_time))

# > time python3 day03.py 
# 0.220s

File: day03/day03.go
package main

import(
	"fmt"
	"os"
	"log"
	"bufio"
	"strconv"
	"strings"
	"math"
)

func Abs(x int) int { // golang doesnt have built-in ab for int64
	if x < 0 {
		return -x
	}
	return x
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
  	}
  	defer file.Close() 

  	type Key struct {
  		X, Y int
  	}
	// seen := map[Key]bool{} // golang doesnt have a set (???)
	seen := map[Key]int{} // Key to minimum steps it took to get there

  	min_distance := math.MaxInt64
  	first := true

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if first {
			line := scanner.Text() // path from central thing
			fmt.Println("first lne")
			fmt.Println(line)
			if err != nil {
				log.Fatal(err)
			}

			counter := 0
			r, c, dr, dc := 0, 0, 0, 0
			dir := strings.Split(line, ",")
			for _, d := range dir {
				if d[0] == 'L' {
					dr, dc = 0, -1
				} else if d[0] == 'R' {
					dr, dc = 0, 1
				} else if d[0] == 'U' {
					dr, dc = 1, 0
				} else if d[0] == 'D' {
					dr, dc = -1, 0 
				}

				steps, err := strconv.Atoi(d[1:])
				if err != nil {
					log.Fatal(err)
				}

				for i := 0; i < steps; i++ {
					r += dr
					c += dc
					counter += 1

					// if this hasnt been visited
					if _, ok := seen[Key{r,c}]; !ok {
						seen[Key{r,c}] = counter
					}
					fmt.Println(Key{r,c})
				}

			}
			fmt.Println(seen)
			first = false
		} else {
			// second line 
			line := scanner.Text()
			
			counter := 0
			r, c, dr, dc := 0, 0, 0, 0
			fmt.Println(line)
			dir := strings.Split(line, ",")
			for _, d := range dir {
				if d[0] == 'L' {
					dr, dc = 0, -1
				} else if d[0] == 'R' {
					dr, dc = 0, 1
				} else if d[0] == 'U' {
					dr, dc = 1, 0
				} else if d[0] == 'D' {
					dr, dc = -1, 0 
				}

				steps, err := strconv.Atoi(d[1:])
				if err != nil {
					log.Fatal(err)
				}
				

				for i := 0; i < steps; i++ {
					r += dr
					c += dc
					counter += 1

					// fmt.Println(Key{r,c})
					if val, ok := seen[Key{r,c}]; ok {
						// manhattan_distance := Abs(r) + Abs(c)
						sum_distance := val + counter
						min_distance = min(sum_distance, min_distance)
					}
				}

			}
		}
	}

	fmt.Println(min_distance)

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}